dynamic resource allocation: API

pohly opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This is the API for [dynamic resource allocation](https://github.com/kubernetes/enhancements/issues/3063), a feature that is targeting alpha in 1.26.

#### Special notes for your reviewer:

For the pull patch series see https://github.com/kubernetes/kubernetes/pull/111023/commits.

The API needs to be agreed upon and (ideally) get merged before adding the other pieces.

#### Does this PR introduce a user-facing change?

Adding releases notes will be done in the PR which completes the feature.

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3063
```

pohly committed:
api: implement printers for dynamic resource allocation

This is needed for "kubectl get". It depends on the generated swagger docs.

pohly committed:
kube features: add DynamicResourceAllocation

pohly commented:
/sig node

pohly commented:
The failures in pull-kubernetes-kind-dra are expected: the feature is not complete yet...

pohly commented:
/cc @thockin

pohly commented:
/assign @thockin

pohly commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
+
+// PodScheduling objects get created by a scheduler when it handles a pod which
+// uses one or more unallocated ResourceClaims with delayed allocation.
+type PodScheduling struct {
```
Perhaps `PodSchedulingState`? The plural form is a bit nicer (`podschedulingstates`).

alculquicondor replied:
PodResourceClaimsScheduling?

pohly replied:
I think this over-emphasizes the use for ResourceClaims. If we ever find some other aspect of a pod which needs this kind of coordination, then the `PodResourceClaimsScheduling` name would become wrong.

pohly replied:
If we put the new type into "dra.k8s.io/v1alpha1", would "SchedulingState" make sense? It would be used as `drav1alpha1.SchedulingState` in Go code.

pohly commented:
/test pull-kubernetes-integration

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
 	Error *string
 }
+
+// PodResourceClaim references exactly one ResourceClaim, either by name or
```
```suggestion
// PodResourceClaim references a ResourceClaim, either by name or
```

just simpler wording

alculquicondor replied:
Actually, the description is confusing, because it's mostly describing ClaimSource, rather than this object.

pohly replied:
True. This comes from a time when ClaimSource didn't exist yet. Updated.

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
+
+// PodResourceClaim references exactly one ResourceClaim, either by name or
+// by embedding a template for a ResourceClaim that will get created
+// by the resource claim controller in kube-controller-manager.
```
```suggestion
// by the control plane.
```

pohly replied:
Done.

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
+	Name string
+
+	// Claim determines where to find the claim.
+	Claim ClaimSource
```
```suggestion
	Source ClaimSource
```

pohly replied:
Works for me. Changed.

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -2997,6 +3000,11 @@ type PodSpec struct {
+	// ResourceClaims defines which ResourceClaims must be allocated
+	// and reserved before the Pod is allowed to start. The resources
+	// will be made available to those containers which reference them
+	// by name.
```
Let's document that this field is immutable.

We are not doing a good job documenting immutability in PodSpec, but let's not make it worse.

pohly replied:
Agreed and done.

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
+	// will *not* be used for the pod to avoid using an unrelated
+	// resource by mistake. Scheduling is then blocked until
+	// the unrelated ResourceClaim is removed. If such a pre-created ResourceClaim is
+	// meant to be used by the pod, the ResourceClaim has to be updated with an
```
I don't think we should make it mutable. The user should create a new pod.

pohly replied:
We cannot prevent the user from manually creating a ResourceClaim that looks like it was created automatically. The API server cannot prevent it because to it a user and the controller look alike (both have permission to create ResourceClaims).

But I guess we don't need to discuss this corner case here. The important part is that the owner reference must be there. I've removed the "If such a pre-created ..." part.

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -2160,6 +2160,9 @@ type ResourceRequirements struct {
 	// +optional
 	Requests ResourceList
+	// The entries are the names of resources in PodSpec.ResourceClaims
+	// that are used by the container.
```
Add a note that this is an alpha-field and which kube_feature controls it. Same for the other new fields.

pohly replied:
Although the [corresponding KEP](https://github.com/kubernetes/enhancements/issues/2558) hasn't made progress? I had checked and it looked like the effort to add that syntax had been abandoned.

Just wondering. I have added it to both internal and v1 types.go, just to be on the safe side. It probably only really needs to be in the external one?

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
+	// this should not be necessary, but it may be useful when
+	// manually reconstructing a broken cluster.
+	//
+	// Running the pod also gets blocked by a wrong ownership. This should
+	// be even less likely because of the prior scheduling check, but could
+	// happen if a user force-deletes or modifies the ResourceClaim.
```
I think we can say in the previous paragraph that scheduling and pod startup are blocked on wrong ownership. Then we can remove this paragraph.

pohly replied:
Agreed.

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
+	// be even less likely because of the prior scheduling check, but could
+	// happen if a user force-deletes or modifies the ResourceClaim.
+	//
+	// This field is read-only and no changes will be made by Kubernetes
+	// to the ResourceClaim after it has been created.
```
This seems unnecessary. The control plane generally doesn't update pod specs.

pohly replied:
What I wanted to say is "the control plane will not update the ResourceClaim because this field is immutable". Updated, I hope it is clearer now.

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
+	// driver deployment.
+	//
+	// The apiserver does not check that the referenced class exists, but a
+	// driver-specific admission webhook may require that and is allowed to
+	// reject claims where the class is missing.
```
that's the case with any webhook, so I don't think we need to document that here.

pohly replied:
Okay.

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
+	// attempt has definitely failed or b) when the allocated resource was
+	// freed. This ensures that resources are not leaked.
+	//
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
```
These are k8s developer docs, not k8s user docs.

pohly replied:
So you want me to remove the link? Done.

alculquicondor replied:
Not sure what you wanted to document, but you could maybe link to the API reference docs? Although I think the generated documentation will have a link to the documentation for ObjectMeta already.

pohly replied:
I think I just wanted to provide more information about ObjectMeta because the type itself doesn't have good documentation. But other types don't do that either, so it should be fine to just do the same also for the new types.

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
+	// disabled.
+	//
+	// The maximum size is 32 (= [ResourceClaimReservedForMaxSize]). This
+	// is an artificial limit to prevent a completely unbounded field in
+	// the API.
```
no need to say this in user docs

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
+	// resource.
+	//
+	// The driver then needs to deallocate this claim and the scheduler
+	// will try again.
```
Is the driver expected to set it back to false? Does ReservedFor need to be empty when setting this to true?

pohly replied:
> Is the driver expected to set it back to false?

Yes. I added an explanation regarding that.

> Does ReservedFor need to be empty when setting this to true?

It wasn't required (deallocation would have been blocked while reserved), but it makes sense to add that because the scheduler should pick some other claim for deallocation or wait. Added.

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
+	// may have been deleted in the meantime.
+	//
+	// The maximum size of this field is 16KiB (= [ResourceHandleMaxSize]).
+	ResourceHandle string
```
should this be an unstructured type?

pohly replied:
You mean `runtime.RawExtension`? That would imply that drivers have to store valid JSON here because the apiserver would validate it. I think an opaque string is better.

pohly replied:
The advantage is that a patch which adds or removes an allocation only needs to touch one field in (ResourceClaimStatus.Allocation). RBAC permissions for drivers can be limited to just the status or even specific fields (once the apiserver supports that).

alculquicondor replied:
Is that true? The driver already needs to add a finalizer

alculquicondor replied:
oh you said "one field", not "one API call".

But changing just one field doesn't sound like a good justification. If it can be done with annotations, it should probably be done with annotations.

pohly replied:
You are right, I had forgotten about the finalizer. So that permission argument is invalid. But the inverse applies: normal users can edit annotations (although they shouldn't) and thus break the driver. In the status, the handle is protected from user mistakes.

Regarding updates, I meant "one API call and one field": in a single status update call it can set the entire "Allocation" field and doesn't need two calls (one for status, one for annotations).

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
+	// Node-local resources can use the `kubernetes.io/hostname` label
+	// to select a specific node.
+	//
+	// Setting this field is optional. If nil, the resource is available
```
```suggestion
	// Setting this field is optional. If null, the resource is available
```

We use the name in YAML in these comments.

pohly replied:
Changed.

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
+
+	// SharedResource determines whether the resource supports more
+	// than one user at a time.
+	SharedResource bool
```
```suggestion
	AllowsMultipleUsers bool
```
Or ShareableResource?

Otherwise it sounds like it shows whether multiple users are connected to it.

pohly replied:
I like `ShareableResource`. Changed.

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
+	// out nodes based on labels. This prevents adding nodes here that the
+	// driver then would need to reject through UnsuitableNodes.
+	//
+	// The size of this field is limited to 256 (=
+	// [PodSchedulingNodeListMaxSize]).  This is large enough for many
+	// clusters. Larger clusters may need more attempts to find a node that
+	// suits all pending resources.
```
It sounds like we should have a test for this for 5k nodes with kubemark. Make sure to sync with sig-scalability when graduating to beta.

pohly replied:
Will do.

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
+// PodSchedulingNodeListMaxSize defines the maximum number of entries in the
+// node lists that are stored in PodScheduling objects. This limit is part
+// of the API.
+const PodSchedulingNodeListMaxSize = 256
```
can we inform this better? How many bytes does this limit imply?

pohly replied:
256 * maximum size of a node name, so 256 * 253 ("DNS1123SubdomainMaxLength is a subdomain's max length in DNS (RFC 1123)") -> ~64KiB.

[According to Tim](https://github.com/pohly/enhancements/pull/13#issuecomment-1163400602), this can be large, it just must not be unlimited.

alculquicondor replied:
Quite a bit already. We capped the limits at 20KiB for other KEPs. Maybe we should do 128.

pohly replied:
Okay, let's do 128 for now. Changed.

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
+	ParametersRef *ResourceClaimParametersReference
+
+	// Allocation can start immediately or when a Pod wants to use the
+	// resource. Waiting for a Pod is the default.
```
```suggestion
	// resource. WaitForFistConsumer is the default.
```

pohly replied:
`AllocationModeWaitForFirstConsumer`? Changed.

pohly replied:
Or literally the value of that constant?

alculquicondor replied:
the value of the constant. That's what's visible to end users

pohly replied:
Fixed.

alculquicondor commented on pkg/apis/core/v1/defaults_test.go:
```diff
@@ -1919,3 +1921,15 @@ func TestSetDefaultServiceInternalTrafficPolicy(t *testing.T) {
+	class := &v1.ResourceClaim{}
+
+	// field should be defaulted
+	defaultMode := v1.AllocationModeWaitForFirstConsumer
```
also it shouldn't change if the field is set

pohly replied:
Added.

alculquicondor reviewed (commented):
First pass. I'm still missing several files.

pohly commented on pkg/api/pod/util_test.go:
```diff
@@ -1935,3 +1935,120 @@ func TestDropHostUsers(t *testing.T) {
+
+	podInfo := []struct {
+		description string
+		hasClaims   bool
+		pod         func() *api.Pod
```
This test is doing it exactly like other tests in this file, see https://github.com/kubernetes/kubernetes/blob/ccf57ba09dafa94906310059e42a6b896c0eaf44/pkg/api/pod/util_test.go#L705-L785

I don't have a strong preference which style is better (compute combinations and outcome vs. explicitly listing all test cases), so I made the change that you suggested in a separate commit. If we prefer consistency or brevity after all, then we can remove it again.

alculquicondor replied:
Existing test code is not necessarily comprehensive or well written, as indicated by this disclaimer https://github.com/kubernetes/enhancements/tree/master/keps/NNNN-kep-template#test-plan

pohly commented on pkg/apis/core/register.go:
```diff
@@ -96,6 +96,12 @@ func addKnownTypes(scheme *runtime.Scheme) error {
 		&RangeAllocation{},
 		&ConfigMap{},
 		&ConfigMapList{},
+		&ResourceClass{},
+		&ResourceClassList{},
+		&ResourceClaim{},
+		&ResourceClaimList{},
+		&PodScheduling{},
+		&PodSchedulingList{},
```
That's what I had initially, but the result was a mixture of some types in core/v1 (ResourceClaimSpec) because they were needed there and other types in a separate apigroup (ResourceClaim). Tim then asked to put everything into core/v1.

Even though these fields and types will appear in core/v1, they will be documented and treated as alpha (not available when feature is off). We cannot add them only in core/v1alpha1 because then they wouldn't survive roundtrips through v1. This is how beta or GA APIs have been extended in the past.

I see that I missed the "alpha" comment in a few places. I'll fix that.

liggitt replied:
> Even though these fields and types will appear in core/v1, they will be documented and treated as alpha (not available when feature is off). We cannot add them only in core/v1alpha1 because then they wouldn't survive roundtrips through v1. This is how beta or GA APIs have been extended in the past.

Adding specific feature-gated *fields* to GA types at different stability levels is a thing we do, but I'm not aware of adding new top-level types to a GA version and characterizing them as alpha. @thockin, is this really what you were asking for?

Also, I thought we didn't want to add new top-level types to the legacy core API group in general.

pohly replied:
FWIW, [Tim wrote](https://github.com/kubernetes/enhancements/pull/3064#discussion_r895081252) during KEP review:

> IIUC, it's exceptionally weird to have half the types in one group and half in another. At that point I'd just argue they should go in core.

That could mean that I need to bring back core/v1alpha1 (which no longer exists) and then promote these types through different stages until they finally land in core/v1. I can do that if that is preferred.

pohly replied:
However, we then have the problem again that documentation for core/v1 shows `ResourceClaimSpec` but not `ResourceClaim`, which might be confusing. it might be simpler and thus less error-prone to just put them into core/v1 right away - just my two cents.

pohly replied:
I think we could still argue about changes while it is marked as alpha, but as soon as we reach beta it would become very hard to make changes. core/v1alpha1 with just the three types would be cleaner, with promotion to core/v1beta1 etc. Is that okay?

alculquicondor replied:
The validation/strategy/etc pretty much LGTM, but this question is very important and it doesn't have a conclusion.

@liggitt @thockin

pohly replied:
FWIW, I am working on moving the three types into "core.k8s.io" v1alpha1.

pohly replied:
I'm not sure yet whether I can use "core.k8s.io" as group name or have to use "" as for the core v1.

pohly replied:
The trend is to use a separate API group after all - see also https://kubernetes.slack.com/archives/C0EG7JC6T/p1667495339738469

pohly commented on pkg/apis/core/fuzzer/fuzzer.go:
```diff
@@ -544,5 +545,25 @@ var Funcs = func(codecs runtimeserializer.CodecFactory) []interface{} {
+
+			// Custom fuzzing for allocation mode.
+			modes := []core.AllocationMode{
+				core.AllocationMode(fmt.Sprintf("%d", c.Rand.Int31())),
```
The default fuzzer then will never hit the valid strings. Other places only set one of the valid strings randomly, for example:
```
			volumeModes := []core.PersistentVolumeMode{core.PersistentVolumeFilesystem, core.PersistentVolumeBlock}
			pv.Spec.VolumeMode = &volumeModes[c.Rand.Intn(len(volumeModes))]
```

I'll do the exact same thing and drop the random allocation mode. I'm not sure anymore why I included it.

alculquicondor commented on pkg/registry/core/resourceclaim/strategy_test.go:
```diff
@@ -0,0 +1,76 @@
+	"k8s.io/kubernetes/pkg/apis/core"
+)
+
+func TestResourceClaimStrategy(t *testing.T) {
```
from experience, testing all functions in a single test make it very hard to write scenarios.

At least split in two:
- one for PrepareForCreate and Validate
- one for PrepareForUpdate and ValidateUpdate

Same for the other strategies.

pohly replied:
I've split all three up into smaller, more logical tests.

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6974,3 +7047,264 @@ func sameLoadBalancerClass(oldService, service *core.Service) bool {
+	// permissions to set this field, kube-scheduler, is a trusted
+	// component.  Also, if it gets this wrong because of a bug, then the
+	// effect is limited (same semantic).
+	if len(spec.PotentialNodes) > core.PodSchedulingNodeListMaxSize {
```
This is intentionally not validating the set aspect because it would be expensive and clients which have permission to write this field are trusted.

alculquicondor replied:
That would be inconsistent with what we do with other resources. For example, in the job status we validate that UIDs for completion tracking are unique, even though the job controller is trusted. Users can always right their own controllers and make mistakes. 128 elements is not a lot (hence why it's important to add limits).

pohly replied:
Okay, will add.

pohly replied:
Done, in a separate commit. I made a note to benchmark this at some point because I am curios what the impact is.

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6974,3 +7047,264 @@ func sameLoadBalancerClass(oldService, service *core.Service) bool {
+	// could be fairly expensive and if a resource driver gets this wrong,
+	// then it is only going to have a negative effect for the pods relying
+	// on this driver.
+	if len(claim.UnsuitableNodes) > core.PodSchedulingNodeListMaxSize {
```
This is intentionally not validating the set aspect because it would be expensive and clients which have permission to write this field are trusted.

Same with node names. If a trusted client puts some string here that isn't a valid node name, nothing bad happens.

If we do all of that additional validation, we don't gain anything and just spend more CPU cycles during updates.

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3096,6 +3116,7 @@ func validateContainerCommon(ctr *core.Container, volumes map[string]core.Volume
 	allErrs = append(allErrs, ValidateVolumeDevices(ctr.VolumeDevices, volMounts, volumes, path.Child("volumeDevices"))...)
+	allErrs = append(allErrs, validateResourceClaimNames(ctr.Resources.Claims, podClaimNames, path.Child("resources", "claims"))...)
 	allErrs = append(allErrs, validatePullPolicy(ctr.ImagePullPolicy, path.Child("imagePullPolicy"))...)
 	allErrs = append(allErrs, ValidateResourceRequirements(&ctr.Resources, path.Child("resources"), opts)...)
```
Sorry, I don't follow. What is "it" here?

alculquicondor replied:
why don't we validate ResrouceClaimNames as part of `ValidateResourceRequirements`?

pohly replied:
Because then I need to push `podClaimNames` even further down the call chain, but you are right, it is the logical place for it -> moved.

pohly commented on pkg/api/pod/util.go:
```diff
@@ -743,3 +744,35 @@ func SeccompAnnotationForField(field *api.SeccompProfile) string {
+		return false
+	}
+
+	return len(podSpec.ResourceClaims) > 0
```
Done.

pohly commented on pkg/api/pod/util_test.go:
```diff
@@ -1935,3 +1935,168 @@ func TestDropHostUsers(t *testing.T) {
+		t.Run(tc.description, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.DynamicResourceAllocation, tc.enabled)()
+
+			oldPod := tc.oldPod()
+			newPod := tc.newPod()
+			wantPod := tc.wantPod()
```
Me neither. It came from the test that I copied, but "when in Rome, do as the Romans do" is not good advice for some areas of the Kubernetes code base...

Changed.

pohly commented on pkg/apis/core/fuzzer/fuzzer.go:
```diff
@@ -544,5 +544,24 @@ var Funcs = func(codecs runtimeserializer.CodecFactory) []interface{} {
+			}
+			obj.AllocationMode = modes[c.Rand.Intn(len(modes))]
+		},
+		func(obj *core.PodScheduling, c fuzz.Continue) {
+			c.FuzzNoCustom(obj) // fuzz self without calling this function again
+		},
+		func(obj *core.PodScheduling, c fuzz.Continue) {
+			c.FuzzNoCustom(obj) // fuzz self without calling this function again
+		},
```
Hmm, good question. Seems to work fine without them -> removed.

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2711,6 +2711,26 @@ func ValidateVolumeDevices(devices []core.VolumeDevice, volmounts map[string]str
 }
 
+func validateResourceClaimNames(claimNames []string, podClaimNames sets.String, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
```
`allErrs := field.ErrorList{}` is all over the place in validation.go files. Would it be worth submitting a PR with a search/replace?

For now I have only fixed in my code and those few functions that I touched.

alculquicondor replied:
yes, that is good enough for this PR

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2711,6 +2711,26 @@ func ValidateVolumeDevices(devices []core.VolumeDevice, volmounts map[string]str
+func validateResourceClaimNames(claimNames []string, podClaimNames sets.String, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	names := sets.NewString()
+	for i, name := range claimNames {
```
I wanted to avoid the allocation during the normal code flow when there are no validation errors and `fldPath.Index(i)` doesn't need to be called. Does that make sense? I can change it if you prefer to avoid the code duplication.

alculquicondor replied:
up to you

pohly commented on pkg/apis/core/types.go:
```diff
@@ -2160,6 +2160,12 @@ type ResourceRequirements struct {
+	// The entries are the names of resources in PodSpec.ResourceClaims
+	// that are used by the container.
+	//
+	// This field is immutable and alpha. It requires enabling the
```
Separate lines probably would get lost during conversion. I went for separate paragraphs:
```
	// This is an alpha field and requires enabling the
	// DynamicResourceAllocation feature gate.
	//
	// It is immutable.
```

pohly commented on pkg/apis/core/types.go:
```diff
@@ -2997,6 +3003,14 @@ type PodSpec struct {
+	// will be made available to those containers which reference them
+	// by name.
+	//
+	// This field is immutable and alpha. It requires enabling the
```
Done.

pohly commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5802,463 @@ type PortStatus struct {
+	AllocationMode AllocationMode
+}
+
+// AllocationMode describes whether a ResourceClaim gets allocated immediately
+// when it gets created (AllocationModeImmediate) or whether allocation is
+// delayed until it is needed for a Pod
+// (AllocationModeWaitForFirstConsumer). Other modes might get added in the
+// future.
```
Done.

pohly replied:
My statement was about the documentation in general. I had a closer look and this particular one, `AllocationMode`, only appears in the go doc output. Therefore I reverted back to referencing the constants.

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3782,6 +3806,55 @@ func validateWindows(spec *core.PodSpec, fldPath *field.Path) field.ErrorList {
+	if claim.Name == "" {
+		allErrs = append(allErrs, field.Required(fldPath.Child("name"), "a name is required to reference the claim"))
+	} else if podClaimNames.Has(claim.Name) {
+		allErrs = append(allErrs, field.Duplicate(fldPath, claim.Name))
```
Fixed.

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3782,6 +3806,55 @@ func validateWindows(spec *core.PodSpec, fldPath *field.Path) field.ErrorList {
+	} else {
+		podClaimNames.Insert(claim.Name)
+	}
+	allErrs = append(allErrs, validatePodResourceClaimSource(claim.Source, fldPath.Child("claim"))...)
```
Fixed.

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6974,3 +7047,264 @@ func sameLoadBalancerClass(oldService, service *core.Service) bool {
+	if ref != nil {
+		if ref.Kind == "" {
+			allErrs = append(allErrs, field.Required(fldPath.Child("kind"), "must be specified"))
+		}
```
It's a reference to arbitrary other objects, therefore it is not known what a well-formed name for those looks like. It's also not possible to check that kind is known because that may change over time (think CRDs).

pohly commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5802,463 @@ type PortStatus struct {
+// ResourceClaimParametersReference contains enough information to let you
+// locate the parameters for a ResourceClaim. The object must be in the same
+// namespace as the ResourceClaim.
+type ResourceClaimParametersReference struct {
```
Tim explicitly asked me to use separate types. My understanding of the rationale is that these types then can be extended independently if needed.

pohly commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5802,463 @@ type PortStatus struct {
+// ResourceClassParametersReference contains enough information to let you
+// locate the parameters for a ResourceClass.
+type ResourceClassParametersReference struct {
+	// APIGroup is the group for the resource being referenced. It is
+	// empty for the core API. This matches the group in the APIVersion
+	// that is used when creating the resources.
+	APIGroup string
+	// Kind is the type of resource being referenced. This is the same
+	// value as in the parameter object's metadata.
+	Kind string
+	// Name is the name of resource being referenced.
+	Name string
```
Embedding would safe some code, but IMHO it would make the API less clear.

pohly commented on pkg/apis/core/types.go:
```diff
@@ -5788,3 +5802,463 @@ type PortStatus struct {
+	Name string
+}
+
+// ResourceClaimParametersReference contains enough information to let you
```
Fixed.

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6974,3 +7047,264 @@ func sameLoadBalancerClass(oldService, service *core.Service) bool {
+}
+
+// validateSet ensures that a slice contains no duplicates and does not exceed a certain maximum size.
+func validateSet[T comparable](set []T, maxSize int, validateItem func(item T, fldPath *field.Path) field.ErrorList, fldPath *field.Path) field.ErrorList {
```
I misread that at first as "is set" as in "is not empty". I switched to `validateIsASet`.

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6974,3 +7047,264 @@ func sameLoadBalancerClass(oldService, service *core.Service) bool {
+	}
+
+	allErrs = append(allErrs, validateAllocationResult(resourceClaim.Status.Allocation, fldPath.Child("allocation"))...)
+	allErrs = append(allErrs, validateSet(resourceClaim.Status.ReservedFor, v1.ResourceClaimReservedForMaxSize,
```
Right.

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6974,3 +7047,264 @@ func sameLoadBalancerClass(oldService, service *core.Service) bool {
+		}
+	}
+
+	if !oldResourceClaim.Status.DeallocationRequested &&
```
Then it is not clear whether the error is caused by `DeallocationRequested` being set to true by the update or whether a new user gets added, i.e. the validation message would have to be more vague.

The other case is handled above where changes to `ReservedFor` are checked, with "reservedFor: new entries may not get added while the claim is meant to be deallocated" as error.

pohly commented on pkg/apis/core/validation/validation_podscheduling_test.go:
```diff
@@ -0,0 +1,337 @@
+			isExpectedFailure: false,
+			scheduling: func() *core.PodScheduling {
+				scheduling := testPodScheduling(goodName, goodNS, goodPodSchedulingSpec)
+				scheduling.DeletionGracePeriodSeconds = &ten
```
Done.

pohly commented on pkg/apis/core/validation/validation_podscheduling_test.go:
```diff
@@ -0,0 +1,337 @@
+	ten := int64(10)
+
+	scenarios := map[string]struct {
+		isExpectedFailure bool
```
But which fields should I exclude? The error details might be a candidate because the explanation for bad labels and names are very long. But if do that unconditionally, then I also loose coverage for cases where I may want to check that the details are right.

I switched to checking all fields. Better thorough than sorry... I used testify/assert, it does diffs and is just a single line.

Looking at the details made me reconsider the decision to report the full `ResourceClaimSchedulingStatus` as duplicated value: that might be too large, the name is enough.

alculquicondor replied:
In general, I at least exclude the value

pohly replied:
Not excluding anything was a useful exercise to ensure that also the values that get reported are useful (for example, not doing a diff in ValidateImmutableField). I think I'll keep it like this.

pohly commented on pkg/apis/core/validation/validation_podscheduling_test.go:
```diff
@@ -0,0 +1,337 @@
+	scenarios := map[string]struct {
+		isExpectedFailure bool
+		oldScheduling     *core.PodScheduling
+		newScheduling     func(scheduling *core.PodScheduling) *core.PodScheduling
```
Done.

pohly commented on pkg/apis/core/validation/validation_podscheduling_test.go:
```diff
@@ -0,0 +1,337 @@
+	validScheduling := testPodScheduling("foo", "ns", core.PodSchedulingSpec{})
+
+	scenarios := map[string]struct {
+		isExpectedFailure bool
```
Done.

pohly commented on pkg/apis/core/validation/validation_podscheduling_test.go:
```diff
@@ -0,0 +1,337 @@
+	validScheduling := testPodScheduling("foo", "ns", core.PodSchedulingSpec{})
+
+	scenarios := map[string]struct {
+		isExpectedFailure bool
+		oldScheduling     *core.PodScheduling
+		newScheduling     func(scheduling *core.PodScheduling) *core.PodScheduling
```
Done.

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6974,3 +7047,264 @@ func sameLoadBalancerClass(oldService, service *core.Service) bool {
+// ValidateResourceClaimUpdate tests if an update to ResourceClaim is valid.
+func ValidateResourceClaimUpdate(resourceClaim, oldResourceClaim *core.ResourceClaim) field.ErrorList {
+	allErrs := ValidateObjectMetaUpdate(&resourceClaim.ObjectMeta, &oldResourceClaim.ObjectMeta, field.NewPath("metadata"))
+	if !apiequality.Semantic.DeepEqual(oldResourceClaim.Spec, resourceClaim.Spec) {
```
Perhaps the Romans were right in this case: ValidateImmutableField reports the entire struct as immutable, without pointing out exactly which field in it was modified.

```
            	            	(field.ErrorList) (len=1) {
            	            	 (*field.Error)({
            	            	  Type: (field.ErrorType) (len=17) "FieldValueInvalid",
            	            	  Field: (string) (len=4) "spec",
            	            	  BadValue: (core.ResourceClaimSpec) {
            	            	   ResourceClassName: (string) (len=5) "valid",
            	            	   ParametersRef: (*core.ResourceClaimParametersReference)({
            	            	    APIGroup: (string) "",
            	            	    Kind: (string) (len=3) "foo",
            	            	    Name: (string) (len=3) "bar"
            	            	   }),
            	            	   AllocationMode: (core.AllocationMode) (len=20) "WaitForFirstConsumer"
            	            	  },
            	            	  Detail: (string) (len=18) "field is immutable"
            	            	 })
```

The solution with `cmp` is more user-friendly. I am going to stick with ValidateImmutableField because it is designed for this purpose. Perhaps it can be enhanced at some later point in time.

alculquicondor replied:
I think we should change the function to use cmp.Diff indeed (but follow up PR)

pohly replied:
https://github.com/kubernetes/kubernetes/issues/113633

pohly commented on pkg/apis/core/validation/validation_resourceclass_test.go:
```diff
@@ -0,0 +1,284 @@
+/*
```
Wise choice :grin: I've updated it the same way as validation_podscheduling_test.go.

alculquicondor commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6974,3 +7048,265 @@ func sameLoadBalancerClass(oldService, service *core.Service) bool {
+	// allocation.
+	if oldResourceClaim.Status.DeallocationRequested &&
+		!resourceClaim.Status.DeallocationRequested &&
+		resourceClaim.Status.Allocation != nil {
```
Can I set allocation to nil without setting deallocationRequested to false? Is that covered?

pohly replied:
It makes sense to prevent that, will add it.

pohly replied:
Done.

I'm also preventing setting deallocationRequested when there is no allocation. That would have been handled by the driver by resetting it, but it makes more sense to not let a claim get into such a state in the first place - it's not needed.

alculquicondor commented on pkg/apis/core/validation/validation_podscheduling_test.go:
```diff
@@ -0,0 +1,312 @@
+		t.Run(name, func(t *testing.T) {
+			scenario.oldScheduling.ResourceVersion = "1"
+			errs := ValidatePodSchedulingUpdate(scenario.update(scenario.oldScheduling.DeepCopy()), scenario.oldScheduling)
+			assert.Equal(t, scenario.wantFailures, errs)
```
I prefer an explicit call to cmp.Diff, but that's just me

pohly committed:
api: add resource claims to core API

The resource.k8s.io/ClaimTemplate only gets referenced by name, therefore the
changes to the core API are limited.

pohly committed:
api: dynamic resource allocation API

This adds a new resource.k8s.io API group with v1alpha1 as version. It contains
four new types: resource.ResourceClaim, resource.ResourceClass, resource.ResourceClaimTemplate, and
resource.PodScheduling.

pohly commented:
As discussed in https://kubernetes.slack.com/archives/C0EG7JC6T/p1667495339738469, we should not add new types directly to core/v1. Adding core/v1alpha1 also doesn't work.

What we can do is add some minimal changes to the core API that are unlikely to be wrong and do all the experimental API in a new v1alpha1 API group. The core API then just references a ClaimTemplate object by name instead of embedding an experimental struct.

After some brainstorming we settled on "resource.k8s.io" as API group. The rationale for that instead of "dra.k8s.io" is:
- more descriptive name
- putting PodScheduling underneath dra is backwards
- allows us to drop the "Resource" prefix of types (resource.Claim, etc.)

I have pushed and update with that change, but need to rebase. The goal is to ensure that all tests pass, get consensus on this API, then rebase the "business logic" from https://github.com/kubernetes/kubernetes/pull/111023 onto the new API. In the meantime, review and updates of that PR with the old API continues.

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6974,3 +7047,264 @@ func sameLoadBalancerClass(oldService, service *core.Service) bool {
+	// component.  Also, if it gets this wrong because of a bug, then the
+	// effect is limited (same semantic).
+	if len(spec.PotentialNodes) > core.PodSchedulingNodeListMaxSize {
+		allErrs = append(allErrs, field.TooLongMaxLength(fldPath.Child("potentialNodes"), nil, core.PodSchedulingNodeListMaxSize))
```
Okay.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2711,6 +2719,71 @@ func ValidateVolumeDevices(devices []core.VolumeDevice, volmounts map[string]str
+	return podClaimNames, allErrs
+}
+
+func validatePodResourceClaim(claim core.PodResourceClaim, podClaimNames sets.String, fldPath *field.Path) field.ErrorList {
```
I know, and you know that `sets.String` is a `map` but we should still pass by pointer so it is clear thsat this func mutates it in the parent context,

pohly replied:
Changed.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2711,6 +2719,71 @@ func ValidateVolumeDevices(devices []core.VolumeDevice, volmounts map[string]str
+func validatePodResourceClaimSource(claimSource core.ClaimSource, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if claimSource.ResourceClaimName != nil && claimSource.TemplateName != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, claimSource, "either resourceClaimName or template must be set, not both"))
```
see other examples for error style, e.g.

```
"may not have more than one field specified"
```

thockin commented on pkg/apis/core/types.go:
```diff
@@ -3000,12 +3017,68 @@ type PodSpec struct {
+	// ResourceClaim.
+	//
+	// Either this or ResourceClaimName must be set, but not both.
+	TemplateName *string
```
ResourceClaimTemplateName

pohly replied:
Done.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2711,6 +2719,71 @@ func ValidateVolumeDevices(devices []core.VolumeDevice, volmounts map[string]str
+	if claimSource.ResourceClaimName != nil && claimSource.TemplateName != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, claimSource, "either resourceClaimName or template must be set, not both"))
+	}
+	if claimSource.ResourceClaimName == nil && claimSource.TemplateName == nil {
```
see other examples for error style, e.g.

```
"must specify one of: `fieldRef`, `resourceFieldRef`, `configMapKeyRef` or `secretKeyRef`"
```

pohly replied:
Changed.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2711,6 +2719,71 @@ func ValidateVolumeDevices(devices []core.VolumeDevice, volmounts map[string]str
+	for i, claim := range claims {
+		name := claim.Name
+		if name == "" {
+			allErrs = append(allErrs, field.Required(fldPath.Index(i), "name cannot be empty"))
```
no last arg to Required

pohly replied:
Removed.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2711,6 +2719,71 @@ func ValidateVolumeDevices(devices []core.VolumeDevice, volmounts map[string]str
+				names.Insert(name)
+			}
+			if !podClaimNames.Has(name) {
+				allErrs = append(allErrs, field.Invalid(fldPath.Index(i), name, "not defined in pod.spec.resourceClaims"))
```
VolumeMounts example is: `allErrs = append(allErrs, field.NotFound(idxPath.Child("name"), mnt.Name))`

pohly replied:
Changed.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2954,7 +3027,7 @@ func validateEphemeralContainers(ephemeralContainers []core.EphemeralContainer,
 
 		c := (*core.Container)(&ec.EphemeralContainerCommon)
-		allErrs = append(allErrs, validateContainerCommon(c, volumes, idxPath, opts)...)
+		allErrs = append(allErrs, validateContainerCommon(c, volumes, podClaimNames, idxPath, opts)...)
```
is it OK to have this in ephemeral containers?  That seems weird to me...

pohly replied:
I could imagine that an ephemeral container may need access to debug something related to the resource, so why not?

thockin commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,466 @@
+//
+// This is an alpha type and requires enabling the DynamicResourceAllocation
+// feature gate.
+type Claim struct {
```
We don't have Go's "don't stutter" rule.  Shouldn't this be `ResourceClaim` ?

pohly replied:
@alculquicondor suggested it and it made sense to me. I can go back to ResourceClaim, ResourceClass, and ResourceClaimTemplate if you prefer that.

pohly replied:
Looks like you really do :grin: 

I'll revert back to the longer names.

thockin replied:
I'm happy to revisit this as a conventions, but we seem to have a general common style of not abbreviating names like this.  Otherwise the YAML `kind` and Go type name don't match and docs get confused.

thockin commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,466 @@
+type Claim struct {
+	metav1.TypeMeta
+
+	// The driver must set a finalizer here before it attempts to allocate
```
I don't think the comments on ObjectMeta will be seen nor do I think it's the right place to document how drivers are supposed to operate?

pohly replied:
verify-description.sh wants some kind of comment here. Putting something useful seemed better than "Standard object metadata" because at least developers will see it when reading the go doc description.

thockin replied:
Requiring a comment on metadata seems dumb to me, but I still don't think this is the place to be instructing driver authors.

sftim replied:
By “driver” my first thought would be “device plugin” - we shouldn't set a puzzle for readers to solve.

pohly replied:
I'll remove all of the ObjectMeta comments. As pointed out above, it's not going to show up in the API documentation and developers should better read the KEP to understand how this works.

thockin commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,466 @@
+	// Spec describes the desired attributes of a resource that then needs
+	// to be allocated. It can only be set once when creating the
+	// Claim.
+	Spec ClaimSpec
```
ResourceClaimSpec?

thockin commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,466 @@
+
+	// Status describes whether the resource is available and with which
+	// attributes.
+	Status ClaimStatus
```
ResourceClaimStatus ?

thockin commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,466 @@
+	// driver deployment.
+	//
+	// The apiserver does not check that the referenced class exists.
+	ClassName string
```
ResourceClassName ?

thockin replied:
I will not repeat, but I THINK youi want all types named very explicitly

thockin commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,466 @@
+	// via the name of a Class that was created as part of the
+	// driver deployment.
+	//
+	// The apiserver does not check that the referenced class exists.
```
not needed - no refs in any type check for existence

pohly replied:
Removed.

thockin commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,466 @@
+// ClaimStatus tracks whether the resource has been allocated and what
+// the resulting attributes are.
+type ClaimStatus struct {
+	// DriverName is a copy of the driver name from the Class at
```
I think the API docs should describe what it is, but not "it's necesssary for..." sorts of things

pohly replied:
Reduced to just the first sentence.

thockin commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,466 @@
+	DriverName string
+
+	// Allocation is set by the resource driver once a resource has been
+	// allocated successfully. Nil indicates that the resource is not
```
"Nil" is a Go-ism that nobody using this API cares about.  "If this is not specified, the resource is not yet allocated."

pohly replied:
Changed.

thockin commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,466 @@
+
+	// ReservedFor indicates which entities are currently allowed to use
+	// the claim. Usually those are Pods, but other objects are also
+	// possible as long as they exist as resources in the apiserver. The
```
We don't really want details of the controllers or where they live - that's all impl details

pohly replied:
Reduced to the first sentence.

thockin commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,466 @@
+	// ReservedFor indicates which entities are currently allowed to use
+	// the claim. Usually those are Pods, but other objects are also
+	// possible as long as they exist as resources in the apiserver. The
+	// Claim controller in the kube-controller-manager checks for
```
I think this needs to be clearer.  Is this right?

"""
If this field specified and points only to objects which do not exist, this claim will be deleted.
"""

?

pohly replied:
Deleting the claim is separate. What the controller does is remove non-existent users from ReservedFor.

But as this is more related to the feature overall and not the API itself, I removed all of this.

thockin commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,466 @@
+	// Claim controller in the kube-controller-manager checks for
+	// stale entries that reference non-existent objects and removes them.
+	//
+	// A scheduler must add a Pod that it is scheduling. This must be done
```
These are sort of impl details - is there not a better place for this?

pohly replied:
The KEP?

I don't think we have any in-tree place where "correct usage of the API" can be described. The API description comes closest, but not if we treat it as "for users".

thockin replied:
@sftim opinions on this?  In particular, should API docs include details about the way that fields are supposed to be used by components like drivers and scheduler?

sftim replied:
> should API docs include details about the way that fields are supposed to be used by components like drivers and scheduler?

I recommend making that a last resort. The kube-controller-manager in particular is written in a way that lets us replace it with, for example, lots of separate components that each handle a different set of API groups. A split isn't likely but - for the API reference - I'm not keen on documenting a particular way of laying out the control plane when Kubernetes' API doesn't actually require that.

We have quite a few precedents for renaming, splitting, or combining controllers. Putting that much detail here feels like an invitation for things to go stale.

thockin commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,466 @@
+	// same Claim, the Claim might have been marked for
+	// deletion, or even been deallocated already.
+	//
+	// kubelet will check this before allowing a Pod to run because a a
```
"a a"

thockin commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,466 @@
+	// same Claim, the Claim might have been marked for
+	// deletion, or even been deallocated already.
+	//
+	// kubelet will check this before allowing a Pod to run because a a
```
Can you make this all clearer.  Pretend to be a user rather than a wizard :)

e.g.

"""
A Pod which references a ResourceClaim which is not reserved for that Pod will not be started.

pohly replied:
I used exactly that sentence because it seems relevant for end-users.

thockin commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,466 @@
+	// doesn't support dynamic resource allocation or the feature was
+	// disabled.
+	//
+	// The maximum size is 32 (= [ClaimReservedForMaxSize]).
```
no need for the symbol name

pohly replied:
Removed.

pohly replied:
... also elsewhere.

thockin commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,466 @@
+}
+
+// ReservedForMaxSize is the maximum number of entries in
+// [ClaimStatus.ReservedFor].
```
This [syntax] is new - are we doing that now?  I don't think so.  Also docs should reference fields by their jsonName not their GoName

pohly replied:
I think this particular type definition will only appear in the go doc output, but to be safe I switched to json names and removed the special go doc formatting.

pohly replied:
It almost feels like we need two versions of these docs: once for end-users, using JSON, and once for Go developers, with all the technical details that they need to correctly work with these APIs and go doc formatting... but not now.

thockin replied:
We need an IDL, but that's a story for another day.

thockin reviewed (commented):
I have to pause for now.  Most of my comments are around docs, so if that is all that is left, I will not block an approval.  More to come ASAP

pohly committed:
api: update testdata

This is the result of

   cd staging/src/k8s.io/api && UPDATE_COMPATIBILITY_FIXTURE_DATA=true go test ./...

pohly commented on pkg/apis/core/types.go:
```diff
@@ -2160,6 +2160,23 @@ type ResourceRequirements struct {
 	// otherwise to an implementation-defined value
 	// +optional
 	Requests ResourceList
+	// The entries are the names of resources in PodSpec.ResourceClaims
```
`spec.resourceClaim` (JSON names), right?

Changed.

pohly commented on pkg/apis/core/types.go:
```diff
@@ -2160,6 +2160,23 @@ type ResourceRequirements struct {
+	// This is an alpha field and requires enabling the
+	// DynamicResourceAllocation feature gate.
+	//
+	// It is immutable.
```
Fixed.

pohly commented on pkg/apis/core/types.go:
```diff
@@ -3000,12 +3017,68 @@ type PodSpec struct {
+	// This is an alpha field and requires enabling the
+	// DynamicResourceAllocation feature gate.
+	//
+	// It is immutable.
```
Fixed.

pohly commented on pkg/apis/core/types.go:
```diff
@@ -3000,12 +3017,68 @@ type PodSpec struct {
+	Source ClaimSource
+}
+
+// ClaimSource references exactly one ResourceClaim, either directly by name or
```
Changed.

pohly commented on pkg/apis/core/types.go:
```diff
@@ -3000,12 +3017,68 @@ type PodSpec struct {
+// Additional options might get added in the future, so code using this struct
+// must error out when none of the options that it supports are set.
+type ClaimSource struct {
+	// The resource is independent of the Pod and defined by
```
Changed.

pohly commented on pkg/apis/core/types.go:
```diff
@@ -3000,12 +3017,68 @@ type PodSpec struct {
+	// the Pod. Either this or Template must be set, but not both.
+	ResourceClaimName *string
+
+	// The name of a dra.k8s.io/ResourceClaimTemplate object in the same namespace.
```
"dra" is out-dated, we went for the more generic "resource.k8s.io" based on @alculquicondor's suggestion.

Let's leave out the apigroup. I don't think it would be linked properly and users of this field need additional documentation to piece it all together anyway.

pohly commented on pkg/apis/core/types.go:
```diff
@@ -3000,12 +3017,68 @@ type PodSpec struct {
+	// the Pod. Either this or Template must be set, but not both.
+	ResourceClaimName *string
+
+	// The name of a dra.k8s.io/ResourceClaimTemplate object in the same namespace.
```
Copied, with some changes:
- ResourceClaimTemplateName
- I left out the exact group.
- I used "resource Claim" and "resource ClaimTemplate" because those are now the type names.

thockin replied:
Can you remind me why this name has to be determinisitic?

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -295,6 +295,14 @@ var ValidateClassName = apimachineryvalidation.NameIsDNSSubdomain
 var ValidatePriorityClassName = apimachineryvalidation.NameIsDNSSubdomain
 
+// ValidateResourceClaimName can be used to check whether the given
+// name for a dra.k8s.io/ResourceClaim is valid.
```
Replaced with "resource Claim" and "resource ClaimTemplate".

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -295,6 +295,14 @@ var ValidateClassName = apimachineryvalidation.NameIsDNSSubdomain
 
+// ValidateResourceClaimName can be used to check whether the given
+// name for a dra.k8s.io/ResourceClaim is valid.
+var ValidateResourceClaimName = apimachineryvalidation.NameIsDNSSubdomain
```
I'm fine with not validating here. Typing an invalid name is not different from typing a valid name of a non-existent object.

Removed.

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2711,6 +2719,71 @@ func ValidateVolumeDevices(devices []core.VolumeDevice, volmounts map[string]str
+func validatePodResourceClaim(claim core.PodResourceClaim, podClaimNames sets.String, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if claim.Name == "" {
+		allErrs = append(allErrs, field.Required(fldPath.Child("name"), "a name is required to reference the claim"))
```
Removed.

pohly committed:
api: fully validate PotentialNodes and SuitableNodes

This is in response to review feedback. Checking for valid node names and the
set property catches programming mistakes in the components that have write
permission.

pohly commented:
I think I have addressed all pending review comments.

@thockin, @alculquicondor: please take another took.

How do you want to merge this? This PR (i.e. just the API), then different PRs for each component, or everything as once (= PR #111023)?

The problem with the monolithic PR is that individual pieces (generated files from the API, go mod files for kubelet dependencies) cause merge conflicts that lead to rebases also of those parts that wouldn't need to be rebased if they were in smaller PRs. The advantage of merging at once is that we don't include a partial solution.

I would prefer to merge separately, but am open to suggestions.

pohly commented on pkg/apis/core/types.go:
```diff
@@ -2160,6 +2160,25 @@ type ResourceRequirements struct {
+}
+
+// ResourceClaim references one entry in PodSpec.ResourceClaims.
+type ResourceClaim struct {
```
I am not happy with the name of this struct anymore. After going back to "ResourceClaim" in resource.k8s.io, there are two structs with this name.

We can still rename it at any time (only affects Go clients, not the API), but perhaps we can already come up with a better name now? 

ResourceClaimRef? That sounds too much like an external reference.

SpecResourceClaim? That's better.

pohly replied:
It is a struct because because we might want to add additional per-entry parameters in the future, similar to mount parameters for volumes.

With "Claims []string" we cannot do that and would have to start using a new field name.

We could do "ClaimNames []string" now and switch to "Claims []ResourceClaimMount" in the future when the need arises, but then we would have to support both fields or switch before going beta. It seemed safer to use the slight more extensible API already now.

Let's not get hung up on the struct name. We can still switch to a nicer name later.

thockin replied:
I agree a different name would make it clearer.  `ResourceClaimMount` is "too cute".  `ResourceClaimAttachment` or `ResourceClaimUser` or `ResourceClaimUse` or similar is fine to me

pohly replied:
But I don't need to change this for alpha because its not visible in the API, right?

thockin replied:
that's fine.  It is visible in API docs, but I can live with it for the moment

pohly commented:
Opens:
- "resource.k8s.io" or "resources.k8s.io" as API group name?
- name for the struct references a spec.ResourceClaims entry by name (https://github.com/kubernetes/kubernetes/pull/112981#discussion_r1014722091)
- how to merge (see above)

thockin commented:
I am a proponent of all-at-once merging, despite the issues raised.  The risk of partial merge or partial rollback is scary.

sftim commented:
@pohly, you stated

> I don't think we have any in-tree place where "correct usage of the API" can be described. The API description comes closest, but not if we treat it as "for users".

We can aim to be much more declarative. For example, if you wanted to state that a field manager asserting control over field _a_ would be expected also to assert control over field _b_, you can state this without naming the most common (or even only known) manager for that field.

It's OK to open a draft PR against the website, /hold that PR because the k/k code is not yet merged, and put a hyperlink to that future documentation into the k/k API docs. Not everyone using Kubernetes has access to the public web, but those folks are a niche audience who can probably work out how to publish a copy of the docs locally.

sftim commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,449 @@
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ResourceClaim is created by users to describe which resources they need.
```
How do we know the creator is a user? Maybe “resource consumer”?

pohly replied:
We can leave out the "created by users" and just focus on what the object is and does -> "ResourceClaim describes which resources are needed."

sftim commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,449 @@
+	// the claim. A Pod which references a ResourceClaim which is not
+	// reserved for that Pod will not be started.
+	//
+	// The maximum size is 32.
```
```suggestion
	// There can be at most 32 such reservations.
```

Are we documenting this as a limit that all implementations of this API must enforce indefinitely, and that clients can rely upon?

pohly replied:
Clients can rely on being able to store this many entries, but not more. This limit may get increased in the future, which should be mentioned.

Various APIs have such limits, usually without documenting them in the API description. They are still part of the API (determined by implementation rather than documentation).

sftim replied:
Mmm, if we document a limit, we're almost implying that this should be part of a conformance test.

We could document the _minimum_ size of the list for a conformant implementation.

pohly replied:
Yes. A conformant implementation must be able to store this many entries (but may store more).

The inverse is also true: a conformant implementation is allowed to reject objects that have more.

sftim commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,449 @@
+	// reserved for that Pod will not be started.
+	//
+	// The maximum size is 32.
+	ReservedFor []ResourceClaimUserReference
```
```suggestion
	ReservedFor []ResourceClaimConsumerReference
```

“user” has a very common meaning in IT. Now we're looking at documenting this, and I'm seeing the potential impact, I think “consumer” would be better.

pohly replied:
Okay.

sftim commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,449 @@
+	// allocated the resource driver to inform the scheduler where it can
+	// schedule Pods using the ResourceClaim.
+	//
+	// Node-local resources can use the `kubernetes.io/hostname` label
+	// to select a specific node.
+	//
```
Could we link to a doc about node selectors instead?

sftim commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,449 @@
+type PodScheduling struct {
+	metav1.TypeMeta
+
+	// The name must be the same as the corresponding Pod.  That Pod must
+	// be listed as owner in OwnerReferences to ensure that the
+	// PodScheduling object gets deleted when no longer needed. Normally
+	// the scheduler will delete it.
```
I wouldn't mention the scheduler here. Instead, detail the expectations on API clients.

pohly replied:
Okay.

pohly replied:
Same with "resource driver" - if we don't call out the role of the scheduler, then nor should we do it for the driver.

sftim commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,449 @@
+	// removed soon.
+	metav1.ObjectMeta
+
+	// Spec is set and updated by the scheduler.
```
I wouldn't mention the scheduler here. Instead, outline what is being specified.

pohly replied:
Updated: "Spec describes where resources for the pod are needed." and "Status describes where resources for the pod can be allocated."

sftim commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,449 @@
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PodScheduling objects get created by a scheduler when it handles a pod which
+// uses one or more unallocated ResourceClaims with delayed allocation.
```
Whilst that's true, the commend should explain what a PodScheduling _is_ first.

pohly replied:
Updated:
```
// PodScheduling objects hold information that is needed to select a node
// for a pod with ResourceClaims that use "WaitForFirstConsumer" allocation
// mode and to trigger that allocation.

sftim commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,449 @@
+// required by a pod and eventually communicates the decision of the scheduler
+// to move ahead with pod scheduling for a specific node.
+type PodSchedulingSpec struct {
+	// When allocation is delayed, the scheduler must set the node for
+	// which it wants the resource(s) to be allocated before the driver(s)
+	// start with allocation.
+	//
+	// The driver must ensure that the allocated resource is available on
+	// this node or update PodSchedulingStatus.UnsuitableNodes to
+	// indicate where allocation might succeed.
+	//
+	// When allocation succeeds, drivers should immediately add the pod to
+	// the ResourceClaimStatus.ReservedFor field together with setting
+	// ResourceClaimStatus.Allocated. This optimization may save scheduling
+	// attempts and roundtrips through the API server because the scheduler
+	// does not need to reserve the claim for the pod itself.
+	//
+	// The selected node may change over time, for example when the initial
+	// choice turns out to be unsuitable after all. Drivers must not
+	// reallocate for a different node when they see such a change because
+	// it would lead to race conditions. Instead, the scheduler will
+	// trigger deallocation of specific claims as needed through the
+	// ResourceClaimStatus.DeallocationRequested field.
```
This should be something like:
```golang
	// For `WaitForFirstConsumer` allocations: the name of the Node where resource
	// allocation is requested.
	//
	// For `Immediate` allocations, ??? 
	SelectedNode string
```

but the existing comment wasn't clear enough about what this should be for the immediate allocation case.

pohly replied:
It's not use for immediate allocation.

A lot of this text described the design, but not this particular field. I reduced the description to:
```
	// SelectedNode is the node for which allocation of ResourceClaims that
	// are referenced by the Pod and that use "WaitForFirstConsumer"
	// allocation is to be attempted.
```

sftim commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,449 @@
+	// ResourceClaimStatus.DeallocationRequested field.
+	SelectedNode string
+
+	// When allocation is delayed, and the scheduler needs to decide on
+	// which node a Pod should run, it will ask the driver(s) on which
+	// nodes the resource might be made available. To trigger that check,
+	// the scheduler provides the names of nodes which might be suitable
+	// for the Pod. Will be updated periodically until all resources are
+	// allocated.
+	//
+	// The ResourceClass.SuiteableNodes node selector can be used to filter
+	// out nodes based on labels. This prevents adding nodes here that the
+	// driver then would need to reject through UnsuitableNodes.
+	//
+	// The size of this field is limited to 128. This is large enough for
+	// many clusters. Larger clusters may need more attempts to find a node
+	// that suits all pending resources.
```
This should be something like:
```golang
	// For `WaitForFirstConsumer` allocations: a subset of the candidate notes where this Pod
	// might be scheduled. The API server may limit the number of entries.
	//
	// For `Immediate` allocations, ??? 
	SelectedNode string
```

but the existing comment wasn't clear enough about what this should be for the immediate allocation case for me to make a suggestion.

Also, `SuiteableNodes` contains a spelling mistake.

pohly replied:
See above.

sftim commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,449 @@
+// PodSchedulingStatus is where resource drivers provide information about
+// where the could allocate a resource and whether allocation failed.
+type PodSchedulingStatus struct {
+	// Each resource driver is responsible for providing information about
+	// those resources in the Pod that the driver manages. It can skip
+	// adding that information when it already allocated the resource.
+	//
+	// A driver must add entries here for all its pending claims, even if
+	// the ResourceSchedulingStatus.UnsuitabeNodes field is empty, because
+	// the scheduler may decide to wait with selecting a node until it has
+	// information from all drivers.
+	ResourceClaims []ResourceClaimSchedulingStatus
+
+	// If there ever is a need to support other kinds of resources
+	// than ResourceClaim, then new fields could get added here
+	// for those other resources.
```
Please rewrite this. The other feedback I've added should give a hint about explaining what these fields are, declaratively. Docs about how to use this API belong elsewhere.

pohly replied:
Done.

sftim commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,449 @@
+	// PodClaimName matches the PodClaim.Name field.
+	PodClaimName string
+
+	// UnsuitableNodes lists nodes that the claim cannot be allocated for.
+	// Nodes listed here will be ignored by the scheduler when selecting a
+	// node for a Pod. All other nodes are potential candidates, either
+	// because no information is available yet or because allocation might
+	// succeed.
+	//
+	// A change of the PodSchedulingSpec.PotentialNodes field and/or a failed
+	// allocation attempt trigger an update of this field: the driver
+	// then checks all nodes listed in PotentialNodes and UnsuitableNodes
+	// and updates UnsuitableNodes.
+	//
+	// It must include the prior UnsuitableNodes in this check because the
+	// scheduler will not list those again in PotentialNodes but they might
+	// still be unsuitable.
+	//
+	// This can change, so the driver also must refresh this information
+	// periodically and/or after changing resource allocation for some
+	// other ResourceClaim until a node gets selected by the scheduler.
+	//
+	// The size of this field is limited to 128, the same as for
+	// PodSchedulingSpec.PotentialNodes.
```
This should be something like:
```golang
	// For `WaitForFirstConsumer` allocations: a subset of the candidate notes where a
	// scheduler should not attempt to place this Pod.
	//
	// For `Immediate` allocations, ???
	//
	// When dynamic resource allocation fails, the failed allocation typically triggers an update
	// to add the name of the failed node into this list.
	// The API server may limit the number of entries in this list (returning a ??? response status
	// if a client requests adding an entry beyond the API server's capacity for entries here).
	// If adding an unsuitable node name fails, the client
	// that attempted the addition should silently ignore that error.
	SelectedNode string
```

but the existing comment wasn't clear enough about what this should be for the immediate allocation case for me to make a suggestion.

Don't mention drivers doing periodic refreshes unless we know that a **watch** verb could not ever work.

pohly replied:
See above.

sftim commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,449 @@
+	// (acme.example.com).
+	DriverName string
+
+	// ParametersRef references an arbitrary separate object that may hold
+	// parameters that will be used by the driver when allocating a
+	// resource that uses this class. The driver will be able to
+	// distinguish between parameters stored here and and those stored in
+	// ResourceClaimSpec. These parameters here can only be set by cluster
+	// administrators.
```
```suggestion
	// ParametersRef references an arbitrary separate object that may hold
	// parameters that will be used by the driver when allocating a
	// resource that uses this class. A dynamic resource driver can
	// distinguish between parameters stored here and and those stored in
	// ResourceClaimSpec.
```

pohly replied:
Done.

sftim commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,449 @@
+	// Standard object metadata
+	metav1.ObjectMeta
+
+	// DriverName determines which resource driver is to be used for
+	// allocation of a ResourceClaim that uses this class.
+	//
+	// Resource drivers have a unique name in reverse domain order
+	// (acme.example.com).
```
```suggestion
	// DriverName defines the name of the dynamic resource driver that is used for
	// allocation of a ResourceClaim that uses this class.
	//
	// Resource drivers have a unique name in reverse domain order
	// (acme.example.com).
```

Aside: why reverse domain order? The rest of Kubernetes usually derives names from domain names in forward order.

pohly replied:
Copied.

Reverse domain order is used for consistency with CSI driver names.

sftim commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,449 @@
+	// administrators.
+	ParametersRef *ResourceClassParametersReference
+
+	// Only nodes matching the selector will be considered by the scheduler
+	// when trying to find a Node that fits a Pod when that Pod uses
+	// a ResourceClaim that has not been allocated yet.
+	//
+	// Setting this field is optional. If null, all nodes are candidates.
```
```suggestion
	// A list of node names. When a scheduler is filtering nodes for placing an incoming
	// Pod, and that Pod uses a dynamic resource from this ResourceClass, the scheduler
	// should treat nodes not on this list as excluded.
	//
	// If this list is empty, a scheduler considers all nodes as potential candidates.
```

Aside: why isn't this either:
- a list of _selectors_?
- part of the `status` of the resource class.

pohly replied:
The intended usage is to select nodes by labels where a tool like NFD sets the labels based on hardware characteristics of the node. The resource driver itself then gets deployed so that it only runs on those nodes.

The set of nodes can change, therefore a list of node names isn't suitable for this purpose. Some controller could update the information in status, but for dynamic selection we have PodScheduling.

sftim reviewed (commented):
I'm afraid this API looks really hard to document.

Here's some partial feedback top illustrate some of my concerns. I'm also going to comment on the enhancement issue.

pohly commented:
I did another push to address @sftim's comments. It has become considerably simpler now.

End-user facing documentation for ResourceClaim/Class/ClaimTemplate will not be too hard. Besides, the hardware vendors will also provide documentation for using their specific drivers.

The biggest complexity is with PodScheduling and status updates, which is relevant for a much smaller audience who (hopefully) will be served well enough by the KEP.

pohly commented:
Can we continue the API review in https://github.com/kubernetes/kubernetes/pull/111023?

That's what we need to get merged.

thockin commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,388 @@
+	// ObjectMeta may contain labels and annotations that will be copied into the PVC
+	// when creating it. No other fields are allowed and will be rejected during
+	// validation.
+	metav1.ObjectMeta
```
Aside: I feel like this is a common enough thing now that we should probably have an ObjectMetaTemplate type that can be used.

thockin reviewed (commented):
Sending some comments now, but still going

pohly commented on pkg/apis/core/types.go:
```diff
@@ -2160,6 +2160,25 @@ type ResourceRequirements struct {
+
+// ResourceClaim references one entry in PodSpec.ResourceClaims.
+type ResourceClaim struct {
+	// Name is the name chosen for a resource claim in PodResourceClaim.
```
Changed.

pohly commented on pkg/apis/core/types.go:
```diff
@@ -3000,12 +3019,66 @@ type PodSpec struct {
+	// This is an alpha field and requires enabling the
+	// DynamicResourceAllocation feature gate.
+	//
+	// This field is immutable..
```
Fixed.

thockin replied:
I still see `..`

pohly replied:
I must have lost my fix somehwere.

Fixed for real over in https://github.com/kubernetes/kubernetes/pull/111023: https://github.com/kubernetes/kubernetes/pull/111023

pohly commented on pkg/apis/core/types.go:
```diff
@@ -3000,12 +3019,66 @@ type PodSpec struct {
+	//
+	// The template will be used to create a new ResourceClaim, which will
+	// be bound to this pod. When this pod is deleted, the ResourceClaim
+	// will also be deleted. The name of the ResourceClaim will be <pod
```
I can do a prototype with that.

pohly replied:
https://github.com/kubernetes/kubernetes/issues/113722

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2711,6 +2711,61 @@ func ValidateVolumeDevices(devices []core.VolumeDevice, volmounts map[string]str
+	return allErrs
+}
+
+func validateResourceClaimNames(claims []core.ResourceClaim, podClaimNames sets.String, fldPath *field.Path) field.ErrorList {
```
Moved.

pohly commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,388 @@
+	// allocated the resource driver to inform the scheduler where it can
+	// schedule Pods using the ResourceClaim.
+	//
+	// Setting this field is optional. If null, the resource is available
```
After the driver is done with the allocation, it knows where the resource is available. For a resource that is reachable via the network, that can be on all nodes. Providing no node selector indicates that.

We cannot use an empty core.NodeSelector for that because that isn't valid content for it (at least one term is required).

thockin replied:
isn't there an optional path where the driver knows the resource is available on a set of machines (or can learn that) but doesn't have a set of labels to represent the set?

pohly replied:
Then the node selector must be "node name" in "set of machine names".

thockin replied:
I'm worried this won't scale, but I am not sure what an answer is other than "set the labels"

sftim replied:
I'm kind of hoping to see a ClusterResourceSlice API or something like that. Similar to EndpointSlices, a Node could appear in more than one ClusterResourceSlice.

If we have a ClusterResourceSlice API, it _could_ define a `nodeSelector` but perhaps would not have to. A similar design to this, basically. A sliced approach feels like it would scale well at the level of what gets persisted and the volume of _writes_, although it might lead to the scheduler starting more watches.

alculquicondor replied:
I think this nodeSelector would have to match a topology domain, to say:
this resource can be used in a particular node, rack, zone, etc.

So yes, I would say the answer is to add the labels.

thockin commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -0,0 +1,325 @@
+
+func validateClaimUserReference(ref resource.ResourceClaimConsumerReference, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	// APIGroup is empty for the legacy group.
```
comment and code do not match?

pohly replied:
Removed the comment, it was out-dated.

pohly commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,388 @@
+
+	// ShareableResource determines whether the resource supports more
+	// than one user at a time.
+	ShareableResource bool
```
Removed.

pohly commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,388 @@
+
+	// ShareableResource determines whether the resource supports more
+	// than one user at a time.
+	ShareableResource bool
```
I had a numeric limit earlier, but the exact semantic was problematic because it wasn't well-defined what it should count - there could be limits for nodes and/or pods.

We decided to postpone that and only do "one consumer" vs. "unlimited". Additional limits can be added later, with "Shareable: true", "MaxPerNode: 5" or something like that.

thockin replied:
Still - does shareable mean "on a node" or "in a cluster" ?

pohly replied:
No limitations whatsoever.

pohly commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,388 @@
+
+	// ShareableResource determines whether the resource supports more
+	// than one user at a time.
+	ShareableResource bool
```
Because it might be different depending on what was allocated.

CSI made the decision to specify some constraints in the CSIDriver object for all volumes of a driver and then found that it was impossible to implement a meta driver that supports different kinds of volumes. Work is under way to replicate the per-driver settings also per-volume.

I also don't think that it would help the scheduler much to know this before allocation. It would still need to wait for allocation to complete, and then it has the information.

thockin commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -0,0 +1,325 @@
+	if driverName == "" {
+		allErrs = append(allErrs, field.Required(fldPath, ""))
+	} else {
+		allErrs = append(allErrs, corevalidation.ValidateQualifiedName(strings.ToLower(driverName), fldPath)...)
```
API comments said reverse-FQDN not a label-style name?

pohly replied:
The comment had it backwards.

pohly commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,388 @@
+// ResourceClaim with "WaitForFirstConsumer" allocation mode.
+type ResourceClaimSchedulingStatus struct {
+	// PodClaimName matches the pod.spec.resourceClaims[*].Name field.
+	PodClaimName string
```
Yes, I guess that would be simpler. I wanted to indicate through the field name that this is not a ResourceClaim.Name, but in the end one has to read the docs to understand that.

I'm changing it.

pohly commented on pkg/apis/resource/types.go:
```diff
@@ -0,0 +1,388 @@
+	// used for allocation of a ResourceClaim that uses this class.
+	//
+	// Resource drivers have a unique name in reverse domain order
+	// (acme.example.com).
```
Indeed. Thanks for pointing it out. The description needs to be updated, the example stays as it is.

pohly replied:
It was at the very beginning and then swapped.

pohly committed:
api: generated files

thockin commented on pkg/registry/resource/podscheduling/storage/storage.go:
```diff
@@ -0,0 +1,138 @@
+		return
+	}
+
+	// TODO: can this and the code above be removed?
```
Yes, the default-on-read stuff is pretty rarely needed, and is usueally a bandaid on existing APIs not new ones

pohly replied:
Removed.

thockin reviewed (commented):
A push was just done, so I need to refresh - some more comments as I push thru

thockin commented on pkg/apis/core/types.go:
```diff
@@ -3000,12 +3019,66 @@ type PodSpec struct {
+
+	// ResourceClaims defines which ResourceClaims must be allocated
+	// and reserved before the Pod is allowed to start. The resources
+	// will be made available to those containers which reference them
```
sounds OK to me ?

pohly replied:
Then I'll change it.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2933,8 +2967,8 @@ func validatePullPolicy(policy core.PullPolicy, fldPath *field.Path) field.Error
 // Note that this is called for pod template even though ephemeral containers aren't allowed in pod templates.
-func validateEphemeralContainers(ephemeralContainers []core.EphemeralContainer, containers, initContainers []core.Container, volumes map[string]core.VolumeSource, fldPath *field.Path, opts PodValidationOptions) field.ErrorList {
-	allErrs := field.ErrorList{}
+func validateEphemeralContainers(ephemeralContainers []core.EphemeralContainer, containers, initContainers []core.Container, volumes map[string]core.VolumeSource, names sets.String, fldPath *field.Path, opts PodValidationOptions) field.ErrorList {
```
changin all of these from `podClaimNames` to `names` is bad - I have no idea what it means in most places.  Names of what?

pohly replied:
Too much sed, not enough brains. I'll bring back the full name.

pohly commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -0,0 +1,325 @@
+	return allErrs
+}
+
+func validateClaimSpec(spec *resource.ResourceClaimSpec, fldPath *field.Path) field.ErrorList {
```
Sigh - it was that before the "remove resource.ResourceClaim stutter" change and then I didn't catch all the places where I had to restore the full name. I'll fix it.

pohly commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -0,0 +1,325 @@
+	}
+	allErrs = append(allErrs, validateClaimParameters(spec.ParametersRef, fldPath.Child("parametersRef"))...)
+	switch spec.AllocationMode {
+	case resource.AllocationModeImmediate, resource.AllocationModeWaitForFirstConsumer:
```
Fixed.

pohly commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -0,0 +1,325 @@
+	// Now check for invariants that must be valid for a ResourceClaim.
+	if len(resourceClaim.Status.ReservedFor) > 0 {
+		if resourceClaim.Status.Allocation == nil {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("reservedFor"), "only allocated claims can get reserved"))
```
Fixed

pohly commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -0,0 +1,325 @@
+				newSet := sets.New(resourceClaim.Status.ReservedFor...)
+				newItems := newSet.Difference(oldSet)
+				if len(newItems) > 0 {
+					allErrs = append(allErrs, field.Forbidden(fldPath.Child("reservedFor"), "new entries may not get added while the claim is meant to be deallocated"))
```
Fixed.

pohly commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -0,0 +1,325 @@
+	if resourceClaim.Status.Allocation == nil &&
+		resourceClaim.Status.DeallocationRequested {
+		if oldClaim.Status.Allocation != nil {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("deallocationRequested"), "must be reset together with removing the allocation"))
```
Yes, but a different one. Without the if/else, the error becomes "deallocationRequested must not be set when the claim is not allocated". That's a true statement, but when the driver was clearing "allocation", the error doesn't make it clear that it should also have cleared "deallocationRequested".

pohly commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -0,0 +1,325 @@
+	if oldClaim.Status.DeallocationRequested &&
+		!resourceClaim.Status.DeallocationRequested &&
+		resourceClaim.Status.Allocation != nil {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("deallocationRequested"), "deallocation cannot be canceled because it may already have started"))
```
Changed to [deallocationRequested] "may only be cleared together with the allocation".

thockin reviewed (commented):
I have read all of this.  I admit some of it is not as deep as I would like, but I did make it throught the API part

pohly commented on pkg/registry/resource/resourceclaim/storage/storage.go:
```diff
@@ -0,0 +1,138 @@
+		return
+	}
+
+	// TODO: can this and the code above be removed?
```
Removed.

klueska commented:
I'm giving my LGTM for the code under `pkg/util/dynamic-resource-allocation/ ` (which will eventually move to a staging repo once https://github.com/kubernetes/org/issues/3821 is done).

The code here is not perfect, but it is acceptable for an `alpha` release in my opinion.
We can revisit and clean it up once this feature moves to beta and we move it under staging.

We will anyway need to iterate on this as the protocol for communicating with the scheduler and other data structures are updated as part of the beta release.

pohly commented:
I switched to keeping https://github.com/kubernetes/kubernetes/pull/111023 up-to-date.

I can force-push the API commits also here if that's desired, but that may include rebases that I need to do because of conflicts in that other PR.

pohly commented:
/close

Merged via https://github.com/kubernetes/enhancements/issues/3063
